### ISSUE TITLE
Windows native library (tensorflow_jni.dll) is dynamically linked to msvcp140.dll

### ISSUE THREAD

**Issue by sabi0** [tensorflow/tensorflow#14456]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 7 SP1 x64
- **TensorFlow installed from (source or binary)**: Downloaded from Maven repository
- **TensorFlow version (use command below)**: 1.4.0

### Describe the problem
I need to run TensorFlow on a Windows PC where I don't have admin rights. So I cannot install MS Visual C++ 2015 Redistributable package. And TensorFlow Windows native library (tensorflow_jni.dll) is dynamically linked to msvcp140.dll from that package.

Thus the only option to run TensorFlow for me is to download the DLL from some untrusted "get any DLL" site and put it somewhere in PATH.

I believe it should not be the case. Please consider dropping dependency on MS Visual C++ 2015 Redistributable package.
Or at least mention this dependency clearly in the "Using TensorFlow with a Maven project" section as well as "Install on Windows" here https://www.tensorflow.org/install/install_java

### Source code / logs
The problem manifests itself as
`Exception in thread "main" java.lang.UnsatisfiedLinkError: C:\Users\...\AppData\Local\Temp\tensorflow_native_libraries-...-0\tensorflow_jni.dll: Can't find dependent libraries`

**Comment by bignamehyp** [comment]
@mrry

**Comment by mrry** [comment]
Asim, I assume you know how the JNI library is built, so can you suggest a solution here?

(Since I don't think it's plausible to remove the dependency on `MSVCP140.dll`, perhaps we need to bundle it with the Maven package?)

**Comment by asimshankar** [comment]
Thanks for the report. I'm not familiar enough with what the convention is when it comes to packaging programs that depend on it.

I'll look into it.

CCing @meteorcloudy @gunan in case they have any thoughts (they may not).

`tensorflow_jni.dll` is built by [this script](https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/tools/ci_build/windows/libtensorflow_cpu.sh#L53), that invokes bazel.  I see that the TensorFlow pip packages for Windows do not package `MSVCP140.dll` either. I'm not sure what the convention is when it comes to distributing it with a program, I'll try to dig into this.

**Comment by sabi0** [comment]
As far as I understand msvcp140.dll is a 'generic' library (iostream, containers, etc.) So it won't be needed if the project was built using MinGW, for example.
Alternatively (if you prefer to stick to MS VC++ compiler) you could probably use `/NODEFAULTLIB` option and supply alternative (statically linked) implementation.

These are my "out of thin air" guesses though. So please ignore them if I don't know what I am talking about.

P.S. Your build script applies some "hacks" with a comment that they can be removed once the following pull request is merged into Bazel:  https://github.com/bazelbuild/bazel/pull/2599
Well, it was merged in March.

**Comment by snnn** [comment]
Hi @sabi0,

I guess your problem is how to get the dlls inside of the "MS Visual C++ 2015 Redistributable" package?  It can be easily solved if you have another computer: Download the package from microsoft.com , install it, then copy the dlls out.  AFAIK, build tensorflow with MinGW toolchains is no longer supported.  Please use VC instead.

**Comment by sabi0** [comment]
Thanks @snnn,

I've already figured out where to get msvcp140.dll
But I believe the problem should be solved in a generic way. So that other users like me don't have to face it in the future.
At the very least "MS Visual C++ 2015 Redistributable" dependency should be clearly mentioned in the System Requirements as well as Installation documentation.

**Comment by snnn** [comment]
Hi @asimshankar 
For how to distribute , this article has some recommendations: 
 https://blogs.msdn.microsoft.com/vcblog/2015/03/03/introducing-the-universal-crt/

**Comment by meteorcloudy** [comment]
Bazel now use /MD to dynamically link to MSVCRT by default, but you can easily switch to static linking(/MT) by adding `--features=static_link_msvcrt` as a build option. (requires Bazel 0.7.0)

But be aware, using `/MT`  might cause problems like https://github.com/tensorflow/tensorflow/issues/12844#issuecomment-328034368

**Comment by sabi0** [comment]
msvcp140.dll is a "Standard C++ Library for native code", not a CTR.
https://msdn.microsoft.com/en-us/library/8kche8ah.aspx

According to MS it is possible (though not recommended) to bundle redistributable DLLs with the application. See https://msdn.microsoft.com/en-us/library/ms235299.aspx

Another option is to statically link the runtime libraries using `/MT` switch. This will likely blow the size of the tensorflow_jni.dll. So you might want to introduce a separate "flavour" of the Maven artifact: `<classifier>static</classifier>`. Keeping the current one (dynamic) as default.

The `/MT` option is described as statically linking _runtime_ libraries. So it might not work for standard C++ library. However it was mentioned as a solution to "MSVCP140.dll missing" question on StackOverflow: https://stackoverflow.com/questions/32998902/msvcp140-dll-missing
So I guess it does work for standard C++ library too.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by meteorcloudy** [comment]
@sabi0 Is this still a problem for you?

**Comment by sabi0** [comment]
I believe, this is a problem "in general", not just for me.
At the very least you should mention in the [installation instructions](https://www.tensorflow.org/install/install_java#install_on_windows) that there is a dependency on MS Visual C++ 2015 Redistributable package.
But an additional artifact (`<classifier>static</classifier>`) would have been a better solution, I believe.

P.S. I've already obtained the msvcp140.dll and placed it into a folder in PATH. So everything works for me.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @asimshankar: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by asimshankar** [comment]
I'm going to go ahead and mark this as "Contributions Welcome" since it doesn't seem we have the required expertise within the TensorFlow team on Windows to work this out.

Help is certainly appreciated. Either in the form of a PR that updates the installation instructions, or ideally in the form of a PR that fixes the bazel BUILD rule. However, before sending any PR for the latter, perhaps describe the plan in a comment in this issue.

Thanks!

**Comment by karankaw** [comment]
I was facing the same Error
java.lang.UnsatisfiedLinkError: tensorflow_native_libraries tensorflow_jni.dll: Can't find depedent libraries

@sabi0 , Thanks for your excellent debugging and RCA , I downloaded & installed Microsoft Visual C++ 2015 Redistributable on Windows Server 2008 R1 machine and it worked for me. 

@asimshankar  Please update The Documentation at [Tensorflow Java Install](https://www.tensorflow.org/install/install_java), atleast it can be mentioned that tensorflow_jni.dll (libtensorflow_jni) depends externally on Microsoft Visual C++ as part of TIPS or Side note.

**Comment by asimshankar** [comment]
@karankaw : That would be great. Could I trouble you (or any other volunteer) to send a PR? (I figured some information from folks who work in Windows more regularly than I do will be better).

The change would be to this file: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/docs_src/install/install_java.md

**Comment by karankaw** [comment]
@asimshankar I have raised PR https://github.com/tensorflow/tensorflow/pull/19719
Please approve, if its OK

**Comment by karankaw** [comment]
@sabi0, I have made some changes wrt mentioning dll dependency to "Install on Windows" section at https://www.tensorflow.org/install/install_java , wchich @asimshankar has approved/merged.
I guess we can close this issue once it gets live on tensorflow site.

**Comment by cogmeta** [comment]
I have followed all the instructions carefully here https://www.tensorflow.org/install/lang_java but i am still getting 

java -cp libtensorflow-1.12.0.jar;. -Djava.library.path=jni HelloTensorFlow
Exception in thread "main" java.lang.UnsatisfiedLinkError: Cannot find TensorFlow native library for OS: windows, architecture: x86_64. See https://github.com/tensorflow/tensorflow/tree/master/tensorflow/java/README.md for possible solutions (such as building the library from source). Additional information on attempts to find the native library can be obtained by adding org.tensorflow.NativeLibrary.DEBUG=1 to the system properties of the JVM.
        at org.tensorflow.NativeLibrary.load(NativeLibrary.java:77)
        at org.tensorflow.TensorFlow.init(TensorFlow.java:66)
        at org.tensorflow.TensorFlow.<clinit>(TensorFlow.java:70)
        at org.tensorflow.Graph.<clinit>(Graph.java:361)
        at HelloTensorFlow.main(HelloTensorFlow.java:9)


@karankaw  can you please help.

**Comment by cogmeta** [comment]
What version of tensorflow have people tried successfully on windows?

**Comment by cogmeta** [comment]
Starting ChromeDriver 2.46.628402 (536cd7adbad73a3783fdc2cab92ab2ba7ec361e1) on port 48819
Only local connections are allowed.
Please protect ports used by ChromeDriver and related test frameworks to prevent access by malicious code.
Mar 04, 2019 10:43:03 AM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: OSS
Exception in thread "Thread-4" java.lang.UnsatisfiedLinkError: C:\Users\olelo\AppData\Local\Temp\tensorflow_native_libraries-1551676423743-0\tensorflow_jni.dll: A dynamic link library (DLL) initialization routine failed
	at java.lang.ClassLoader$NativeLibrary.load(Native Method)
	at java.lang.ClassLoader.loadLibrary0(Unknown Source)
	at java.lang.ClassLoader.loadLibrary(Unknown Source)
	at java.lang.Runtime.load0(Unknown Source)
	at java.lang.System.load(Unknown Source)
	at org.tensorflow.NativeLibrary.load(NativeLibrary.java:101)
	at org.tensorflow.TensorFlow.init(TensorFlow.java:66)
	at org.tensorflow.TensorFlow.<clinit>(TensorFlow.java:70)
	at org.tensorflow.Graph.<clinit>(Graph.java:361)
	at ai.olelo.automation.OleloSimpleWebAutomator.runOlelo(OleloSimpleWebAutomator.java:707)
	at ai.olelo.automation.OleloSimpleWebAutomator$4.run(OleloSimpleWebAutomator.java:879)

**Comment by pacher** [comment]
@sabi0 How did you figure out that it is msvcp140.dll which is missing?
I am trying to run tensorflow on gpu and getting this generic `tensorflow_jni.dll: Can't find dependent libraries` error. How can I know what it missing?
Thanks in advance ;-)

**Comment by sabi0** [comment]
@patcher You need a tool like dumpbin or DLL Export Viewer to list the DLL imported functions.

Here is what I have for tensorflow_jni.dll 1.5.0:
```
ADVAPI32.dll
api-ms-win-crt-convert-l1-1-0.dll
api-ms-win-crt-environment-l1-1-0.dll
api-ms-win-crt-filesystem-l1-1-0.dll
api-ms-win-crt-heap-l1-1-0.dll
api-ms-win-crt-locale-l1-1-0.dll
api-ms-win-crt-math-l1-1-0.dll
api-ms-win-crt-runtime-l1-1-0.dll
api-ms-win-crt-stdio-l1-1-0.dll
api-ms-win-crt-string-l1-1-0.dll
api-ms-win-crt-time-l1-1-0.dll
api-ms-win-crt-utility-l1-1-0.dll
KERNEL32.dll
MSVCP140.dll
SHLWAPI.dll
VCRUNTIME140.dll
```

**Comment by pacher** [comment]
Thanks a lot, I have similar picture in https://github.com/lucasg/Dependencies
Everything seems fine, yet `Can't find dependent libraries` error.
My problem is actually only with GPU, like here https://github.com/tensorflow/tensorflow/issues/23656
In that issue the solution is to update python... But I don't use python, I am trying to run inference in java ¯\_(ツ)_/¯

**Comment by sabi0** [comment]
Did you check the 2nd level imports? (What's imported by MSVCP140.dll, etc.)
It is also important that JVM architecture (32 or 64 bits) matches that of loaded libraries. I.e. you cannot load a 64-bit DLL into a 32-bit JVM process and vice versa.

**Comment by pacher** [comment]
Yes, but it is a lot. Since I have no experience with windows native development it gets overwhelming and does not tell me much.
Dependency Walker says that there are some deps missing, like API-MS-WIN-*, EXT-MS-WIN-* and IESHIMS.dll, but it is normal as I understood.
I know about architectures, but I am not doing anything manually, only include tensorflow from maven in my gradle build. Anyway it must be doing the right thing since CPU version starts and seems to work. Unfortunately the network I've got is using tensor format that works only on GPU. And after swapping to `libtensorflow_jni_gpu` I get this `Can't find dependent libraries` error. Must be something with CUDA and friends then. I was hoping there is an easy way to figure out what exactly is missing ;-)

**Comment by swzaaaaaaa** [comment]
> Yes, but it is a lot. Since I have no experience with windows native development it gets overwhelming and does not tell me much.
> Dependency Walker says that there are some deps missing, like API-MS-WIN-_, EXT-MS-WIN-_ and IESHIMS.dll, but it is normal as I understood.
> I know about architectures, but I am not doing anything manually, only include tensorflow from maven in my gradle build. Anyway it must be doing the right thing since CPU version starts and seems to work. Unfortunately the network I've got is using tensor format that works only on GPU. And after swapping to `libtensorflow_jni_gpu` I get this `Can't find dependent libraries` error. Must be something with CUDA and friends then. I was hoping there is an easy way to figure out what exactly is missing ;-)

Hello, I have the same problem as you.On windows10 systems, the call to the tensorflow-cpu version using Java runs successfully, but replacing libtensorflow_jni with libtensorflow_jni_gpu will report an error in Can't find dependent libraries.Excuse me, have you solved it?

**Comment by swzaaaaaaa** [comment]
@pacher Hello, I have the same problem as you.On windows10 systems, the call to the tensorflow-cpu version using Java runs successfully, but replacing libtensorflow_jni with libtensorflow_jni_gpu will report an error in Can't find dependent libraries.Excuse me, have you solved it?

**Comment by pacher** [comment]
@swzaaaaaaa No, unfortunately not. It seems that the only way to get working tensorflow gpu on windows is to install it with Anaconda. So I ended up calling it through python using [Jep](https://github.com/ninia/jep). Jep is awesome, but to make it work on windows is also quite a challenge. It needs to compile native bits during installation using exactly the same C++ compiler as your python was compiled with. After lots of trial and error I succeed, but sadly forgot to write the exact steps down. Now I have a working installation which I am afraid to touch or update. 

From my notes back then, maybe it helps:

> Jep install is easy: `pip install jep`, but you need RIGHT C++ compiler available.
> On linux it's enough to `apt install build-essential` and `apt install python3-dev`
> But on Windows it is quite a different story....
> For python 3.6 you need Visual c++ 2015(!) command line build tools
> This tools are even not visible on M$ download site
> Working thing could be downloaded from:
>      https://download.microsoft.com/download/5/f/7/5f7acaeb-8363-451f-9425-68a90f98b238/visualcppbuildtools_full.exe?fixForIE=.exe
> or   https://download.microsoft.com/download/E/E/D/EEDF18A8-4AED-4CE0-BEBE-70A83094FC5A/BuildTools_Full.exe
> but probably the first link
> Then to install jep you have to:
>      1. use one of the "Visual C++ 2015" (don't remember which one) shortcuts in start menu
> to start command prompt
>      2. From this prompt activate conda environment
>      3. Finally `pip install jep`

> And for python 3.7 you will probably need Visual C++ 2017...

**Comment by swzaaaaaaa** [comment]
@pacher Thank you for your answer, but I have a question, you installed jep, is the language from Java back to python, so how to connect with javaweb?
